File: CS/XPO/CustomEventsAndResources.Module/BusinessObjects/ApplicationUser.cs
using System.ComponentModel;
using System.Drawing;
using System.Text;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Security;
using DevExpress.Persistent.Base.General;
using DevExpress.Persistent.BaseImpl.PermissionPolicy;
using DevExpress.Xpo;
using DomainComponents.Common;

namespace CustomEventsAndResources.Module.BusinessObjects;

[MapInheritance(MapInheritanceType.ParentTable)]
[DefaultProperty(nameof(UserName))]
public class ApplicationUser : PermissionPolicyUser, ISecurityUserWithLoginInfo, IResource {
    public ApplicationUser(Session session) : base(session) { }

    [Browsable(false)]
    [Aggregated, Association("User-LoginInfo")]
    public XPCollection<ApplicationUserLoginInfo> LoginInfo {
        get { return GetCollection<ApplicationUserLoginInfo>(nameof(LoginInfo)); }
    }

    IEnumerable<ISecurityUserLoginInfo> IOAuthSecurityUser.UserLogins => LoginInfo.OfType<ISecurityUserLoginInfo>();

    ISecurityUserLoginInfo ISecurityUserWithLoginInfo.CreateUserLoginInfo(string loginProviderName, string providerUserKey) {
        ApplicationUserLoginInfo result = new ApplicationUserLoginInfo(Session);
        result.LoginProviderName = loginProviderName;
        result.ProviderUserKey = providerUserKey;
        result.User = this;
        return result;
    }
    #region Resource
    public override void AfterConstruction() {
        base.AfterConstruction();
        Color = Color.White;
    }
    [Browsable(false)]
    public object Id {
        get => Oid;
    }
    [NonPersistent]
    public string Caption { get => UserName; set => UserName = value; }
    [Browsable(false)]
    public Int32 OleColor {
        get { return ColorTranslator.ToOle(Color); }
    }
    [Association("CustomEventWithUserResources-ApplicationUser")]
    public XPCollection<CustomEventWithUserResources> Events {
        get { return GetCollection<CustomEventWithUserResources>(nameof(Events)); }
    }
    [ValueConverter(typeof(ColorToIntConverter))]
    public Color Color {
        get => GetPropertyValue<Color>(nameof(Color));
        set => SetPropertyValue(nameof(Color), value);
    }
    #endregion
}

File: VB/XPO/CustomEventsAndResources.Module/BusinessObjects/ApplicationUser.vb
Imports System.ComponentModel
Imports System.Drawing
Imports DevExpress.ExpressApp.Security
Imports DevExpress.Persistent.Base.General
Imports DevExpress.Persistent.BaseImpl.PermissionPolicy
Imports DevExpress.Xpo
Imports DomainComponents.Common

Namespace CustomEventsAndResources.Module.BusinessObjects

    <MapInheritance(MapInheritanceType.ParentTable)>
    <DefaultProperty(NameOf(PermissionPolicyUser.UserName))>
    Public Class ApplicationUser
        Inherits PermissionPolicyUser
        Implements ISecurityUserWithLoginInfo, IResource

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        <Browsable(False)>
        <Aggregated, Association("User-LoginInfo")>
        Public ReadOnly Property LoginInfo As XPCollection(Of ApplicationUserLoginInfo)
            Get
                Return GetCollection(Of ApplicationUserLoginInfo)(NameOf(ApplicationUser.LoginInfo))
            End Get
        End Property

        Private ReadOnly Property UserLogins As IEnumerable(Of ISecurityUserLoginInfo)
            Get
                Return LoginInfo.OfType(Of ISecurityUserLoginInfo)()
            End Get
        End Property

        Private Function CreateUserLoginInfo(ByVal loginProviderName As String, ByVal providerUserKey As String) As ISecurityUserLoginInfo Implements ISecurityUserWithLoginInfo.CreateUserLoginInfo
            Dim result As ApplicationUserLoginInfo = New ApplicationUserLoginInfo(Session)
            result.LoginProviderName = loginProviderName
            result.ProviderUserKey = providerUserKey
            result.UserProp = Me
            Return result
        End Function

#Region "Resource"
        Public Overrides Sub AfterConstruction()
            MyBase.AfterConstruction()
            Color = Color.White
        End Sub

        <Browsable(False)>
        Public ReadOnly Property Id As Object Implements IResource.Id
            Get
                Return Me.Oid
            End Get
        End Property

        <NonPersistent>
        Public Property Caption As String Implements IResource.Caption
            Get
                Return UserName
            End Get

            Set(ByVal value As String)
                UserName = value
            End Set
        End Property

        <Browsable(False)>
        Public ReadOnly Property OleColor As Int32
            Get
                Return ColorTranslator.ToOle(Color)
            End Get
        End Property

        <Association("CustomEventWithUserResources-ApplicationUser")>
        Public ReadOnly Property Events As XPCollection(Of CustomEventWithUserResources)
            Get
                Return GetCollection(Of CustomEventWithUserResources)(NameOf(ApplicationUser.Events))
            End Get
        End Property

        <ValueConverter(GetType(ColorToIntConverter))>
        Public Property Color As Color
            Get
                Return GetPropertyValue(Of Color)(NameOf(ApplicationUser.Color))
            End Get

            Set(ByVal value As Color)
                SetPropertyValue(NameOf(ApplicationUser.Color), value)
            End Set
        End Property
#End Region
    End Class
End Namespace
